adds new script for administrator's reference regarding OAuth2 permission grants in Entra ID (#6536)
pnp/cli-microsoft365: docs/docs/sample-scripts/entra/entra-remove-oauth2-permission-grants/index.mdx

The script first defines a function `Remove-UserPermissionsGrants` that does the necessary work described above and then calls this function with a predefined app registration name.

<Tabs>
  <TabItem value="PowerShell">

  ```powershell
  .SYNOPSIS
    Removes OAuth2 permission grants for a specified Entra ID application registration.

  .DESCRIPTION
    This function removes OAuth2 permission grants (delegated permissions) from an Entra ID application registration.
    It can filter and remove permissions based on how they were granted - either by admin consent, user consent, or both.

  .PARAMETER appRegistrationName
    The display name of the application registration in Entra ID from which to remove permission grants.
    This parameter is mandatory.

  .PARAMETER GrantType
    Specifies which type of permission grants to remove. Valid values are:
    - 'By Admin': Removes only permissions granted through admin consent (consentType = AllPrincipals)
    - 'By User': Removes only permissions granted through user consent (consentType != AllPrincipals)
    - 'All': Removes both admin and user consented permissions
    Default value is: 'Granted By User'

  .EXAMPLE

>     Remove-UserPermissionsGrants -appRegistrationName "My App"

Reviewer: Let's correct the function name here, or correct the actual function name

    Removes all user-consented permission grants for the application "My App"

  .EXAMPLE
    Remove-UserPermissionsGrants -appRegistrationName "My App" -GrantType "All"
    Removes all permission grants (both admin and user consented) for the application "My App"

  .EXAMPLE
    Remove-UserPermissionsGrants -appRegistrationName "My App" -GrantType "By Admin"
    Removes only admin-consented permission grants for the application "My App"

  .NOTES
    Prerequisites:
    - Microsoft 365 CLI (m365) must be installed and configured
    - User must have appropriate permissions in Entra ID to manage OAuth2 permission grants (e.g. Global Admin)

  .LINK
    https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/manage-application-permissions
  #>

  function Remove-PermissionsGrants {
    param(
      [Parameter(Mandatory = $true)]
      [string]$appRegistrationName,

      [Parameter(Mandatory = $false)]

>       [ValidateSet('By Admin', 'By User', 'All')] $GrantType = 'By User'

Reviewer: Let's use similar casing for both parameters. Either camel case or pascal case

Author: Changed it to Pascal Case

    )
    
    Clear-Host
    Write-Host "Removing permission grants for the app registration '$appRegistrationName'"
    Write-Host "Selected grant type: '$GrantType'`n"

    # Get Service Principal that represents the app registration

>     $app = m365 entra enterpriseapp list --query "[?appDisplayName == '$($appRegistrationName)']" | ConvertFrom-Json

Reviewer: Why don't we use `m365 entra enterpriseapp get --displayName $appRegistrationName` here?
Why execute a list command where we can a get command?

Author: Absolutely right. I've made the script more resilient by using the `list` command with a filter. From a performance perspective, the `get` command would be more efficient.
I therefore change this sequence according to your suggestion with an appropriate error handling.

    # Filter (delegated) permissions granted with according consent for the given service principal
    $selectedPermissionsGrants = @()
    if ($GrantType -eq 'All') {

>       $selectedPermissionsGrants = m365 entra oauth2grant list --spObjectId $app.id | ConvertFrom-Json

Reviewer: My output default is currently `text`, which meant that this function errored out because it expects json. Let's add --output json to all commands to make sure cli settings do not break the execution..

Author: Good catch! Sure thing, I'll add `--output json` to all commands. I didn't think of that.

    }

>     elseif ($GrantType -eq 'Granted By Admin') {

Reviewer: I think this should read `By Admin`, as that's the value that's coming through...

      $selectedPermissionsGrants = m365 entra oauth2grant list --spObjectId $app.id --query "[?consentType == 'AllPrincipals']" | ConvertFrom-Json
    }
    else {
      # granted by user
      $selectedPermissionsGrants = m365 entra oauth2grant list --spObjectId $app.id --query "[?consentType != 'AllPrincipals']" | ConvertFrom-Json
    }

>     Write-Host "👉 $($selectedPermissionsGrants.Length) definitions found"

Reviewer: Let's write an empty line below this one, it makes for better readability...

screenshot from my end:

    
    # Remove the filtered permissions
    $selectedPermissionsGrants | ForEach-Object {

>       Write-Host "permission grant (ID) '$($_.id)': " -NoNewline

Reviewer: ```suggestion
      Write-Host "Removing permission grant (ID) '$($_.id)': " -NoNewline
```

Reviewer: Either write Removing before the message, or change `done` to `removed`.

      m365 entra oauth2grant remove --grantId $_.id --force
      Write-Host "✔ Done" -ForegroundColor Green
      Write-Host "⎿ " $_.scope.trim()
    }

    # Display success message
    if($selectedPermissionsGrants.Length -gt 0) {
      Write-Host "`n✔️ Successfully removed selected permissions" -ForegroundColor Green
    }
  }

  # Set the app registration name for which to remove user consents
  $appRegistrationName = 'My App'

>   Remove-UserPermissionsGrants -appRegistrationName $appRegistrationName

Reviewer: Shouldn't this be `Remove-PermissionsGrants`, like the actual function?

    The display name of the enterprise application or service princial name in Entra ID from which to remove permission grants.
    This parameter is mandatory.

  .PARAMETER GrantType
    Specifies which type of permission grants to remove. Valid values are:
    - 'ByAdmin': Removes only permissions granted through admin consent (consentType = AllPrincipals)
    - 'ByUser': Removes only permissions granted through user consent (consentType != AllPrincipals)
    - 'All': Removes both admin and user consented permissions
    Default value is: 'ByUser'

  .EXAMPLE
    Remove-PermissionsGrants -enterpriseAppName "My App"
    Removes all user-consented permission grants for the enterprise application "My App"

  .EXAMPLE
    Remove-PermissionsGrants -enterpriseAppName "My App" -GrantType "All"
    Removes all permission grants (both admin and user consented) for the enterprise application "My App"

  .EXAMPLE
    Remove-PermissionsGrants -enterpriseAppName "My App" -GrantType "ByAdmin"
    Removes only admin-consented permission grants for the enterprise application "My App"

  .NOTES
    Prerequisites:
    - Microsoft 365 CLI (m365) must be installed and configured

>     - User must have appropriate permissions in Entra ID to manage OAuth2 permission grants (e.g. Global Admin)

Reviewer: I think the role "Cloud Application Administrator" can also be used. Let's mention that one as well to show you dont always need god mode.

And let's add one extra prerequisite:

```text
- The Entra app registration used to sign into the CLI for Microsoft 365 must have an appropriate permission scope to manage permissions of other apps (e.g. Directory.ReadWrite.All).
```

You could also try to find out the least permissive scope to use here. I don't know it by heart, but Directory.ReadWrite.All is a bit massive for this.

Author: According to official documentation (Microsoft learn), this would be least permissive Graph scopes to…
- list oauth2PermissionGrants: `Directory.Read.All` (https://learn.microsoft.com/en-us/graph/api/oauth2permissiongrant-list?view=graph-rest-1.0&tabs=http#permissions)
- delete oauth2PermissionGrants: `DelegatedPermissionGrant.ReadWrite.All` (https://learn.microsoft.com/en-us/graph/api/oauth2permissiongrant-delete?view=graph-rest-1.0&tabs=http#permissions)

That's why I'd suggest using `DelegatedPermissionGrant.ReadWrite.All`.

Author: ```suggestion
    - The Entra app registration used to sign into the CLI for Microsoft 365 must have an appropriate permission scope to manage permissions of other apps (at least `DelegatedPermissionGrant.ReadWrite.All`)
```

Reviewer: Yes, that seems the right one. Awesome!

  .LINK

>     https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/manage-application-permissions

Author: Changed as requested

  #>

  function Remove-PermissionsGrants {
    param(
      [Parameter(Mandatory = $true)]

>       [string]$enterpriseAppName,

Reviewer: Let's capitalize this one, just like GrantType.

```suggestion
      [string]$EnterpriseAppName,
```
You could also rename it to AppDisplayName. It's not really important here that it is called an 'enterprise' application.

Author: Changed as requested
